nemo_skills/inference/server/model.py

```diff
@@ -493,10 +493,200 @@ def _send_request(
+
+    def _single_call(
+        self,
+        prompts,
+        tokens_to_generate,
+        temperature,
+        top_p,
+        top_k,
+        repetition_penalty,
+        random_seed,
+        stop_phrases: List[str],
+    ):
+        request = {
+            'prompt': prompts,
+            'max_tokens': tokens_to_generate,
+            'temperature': temperature,
+            'top_p': top_p,
+            'top_k': top_k,
+            'num_generations': 1,  # VLLM provides 1 generation per prompt, duplicate prompts if you want more
+            'stop': stop_phrases,
+            'echo': False,
+            'repetition_penalty': repetition_penalty,
+            'frequency_penalty': 0.0,
+            'presence_penalty': 0.0,
+            'logprobs': None,
+            'logit_bias': None,
+            'seed': random_seed,
+        }
+
+        return self._send_request(request)
+
+    def _send_request(self, request):
+        if self.ssh_server and self.ssh_key_path:
+            raise NotImplementedError("SSH tunnelling is not implemented for vLLM OpenAI server.")
+
+        # temperature of 0 means greedy, but it's not always supported by the server
+        # so setting explicit greedy parameters instead
+        if request["temperature"] == 0:
+            request["temperature"] = 1.0
+            request["top_k"] = 1
+            request["top_p"] = 1.0
+
+        outputs = self.prompt_api(
+            prompt=request['prompt'],
+            max_tokens=request['max_tokens'],
+            temperature=request['temperature'],
+            top_p=request['top_p'],
+            top_k=request['top_k'],
+            num_generations=request.get('num_generations', 1),
+            stop=request['stop'],
+            echo=request.get('echo', False),
+            repetition_penalty=request.get('repetition_penalty', 1.0),
+            frequency_penalty=request.get('frequency_penalty', 0.0),
+            presence_penalty=request.get('presence_penalty', 0.0),
+            logprobs=request.get('logprobs', None),
+            logit_bias=request.get('logit_bias', None),
+            seed=request.get('seed', None),
+            parse_response=True,
+        )
+
+        return outputs
+
+    def prompt_api(
+        self,
+        prompt: str,
+        max_tokens: int,
+        temperature: float,
+        top_p: float,
+        top_k: int = 1,
+        num_generations: int = 1,
+        stop=None,
+        echo: bool = False,
+        repetition_penalty: float = 1.0,
+        frequency_penalty: float = 0.0,
+        presence_penalty: float = 0.0,
+        logprobs: int = None,
+        logit_bias: dict = None,
+        seed: int = None,
+        parse_response: bool = True,
+    ) -> Union["openai.types.Completion", List[str]]:
+
+        if isinstance(prompt, (list, tuple)):
+            prompt = [str(prmt) for prmt in prompt]
+
+        if temperature == 0.0:
+            temperature = 1.0
```

Comment:
This should be done alongside setting top_k to 1 always? Otherwise you risk accidentally setting temp=1 when it was not intended to be so, if for some reason you get a request with t=0, top_k=32, e.g.